### PR TITLE
Cache the `id` attribute on Element

### LINKED ISSUE #6359: Cache the ID in a field on Element to optimize selectors::Element::get_id
Right now, we loop over all the attributes, which is probably too slow.

Code is in `components/script/dom/element.rs`.

### PR DESCRIPTION
**Pull request by asabil**
First stab at contributing to servo. Tried to fix #6359

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6717)

### CODE REVIEW

**Comment by asabil** [comment]
I was not sure if/how layout/wrapper.rs should also be updated to make use of the id field on Element?

**Comment by nox** [comment]
I reviewed what you currently pushed, there is just a small change to do.

For layout/wrapper.rs, you should add a method get_id_for_layout() in the RawLayoutElementHelpers trait and use it there. See get_checked_state_for_layout() for an example of how such a method should look like.

+S-needs-code-changes 

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/dom/element.rs, line 1513 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6717#-JuwkcxKhK0TsKmtOdPm)** ([raw file](https://github.com/servo/servo/blob/ef1a59121b4c12b0d06781a393c3e6a3130d6936/components/script/dom/element.rs#L1513)):</sup>
Don't map, use unwrap() instead (see below in the "if !value.is_empty() { … }" block). We have an id attribute, so if its value isn't an atom, there is something very wrong somewhere. You can see in the virtual method parse_plain_attribute() in this very file how id attributes always end up as an atom.

Actually, you could use expect() here with the panic message that you removed from get_id().

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6717)

**Comment by Ms2ger** [comment]
No! Don't add anything to `RawLayoutElementHelpers`, I'm trying to remove it.

**Comment by nox** [comment]
Sorry, it should be added on LayoutElementHelpers, see local_name() in that very trait. Repeating this from IRC for book-keeping.

**Comment by asabil** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/dom/element.rs, line 1513 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6717#-JuwkcxKhK0TsKmtOdPm)** ([raw file](https://github.com/servo/servo/blob/ef1a59121b4c12b0d06781a393c3e6a3130d6936/components/script/dom/element.rs#L1513)):</sup>
As discussed on irc, not sure exactly what is the correct behaviour here regarding the empty string as an id? /cc @Ms2ger

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6717)

**Comment by jdm** [comment]
Is a DOMRefCell required, or can we use a regular Cell here?

**Comment by nox** [comment]
@jdm I was wondering that too but Cell needs Copy. Atom isn't Copy (or I can't read, which wouldn't be the first time tonight).

**Comment by metajack** [comment]
pinging @Ms2ger

File: components/script/dom/element.rs [hunk 1]
```diff
@@ -1664,12 +1670,7 @@ impl<'a> ::selectors::Element for &'a Element {
         node.get_focus_state()
     }
     fn get_id(&self) -> Option<Atom> {
-        self.get_attribute(&ns!(""), &atom!("id")).map(|attr| {
-            match *attr.r().value() {
-                AttrValue::Atom(ref val) => val.clone(),
-                _ => panic!("`id` attribute should be AttrValue::Atom"),
-            }
-        })
+        self.id_attribute.borrow().as_ref().map(|id| id.clone())
```
Comment (Ms2ger) [inline components/script/dom/element.rs]: I think this can just be `.borrow().clone()`.

**Comment by Ms2ger** [comment]
@asabil sorry we let this slip. I'm happy to land this with or without the nitpick I just posted; please let me know if you can fix it?

**Comment by asabil** [comment]
@Ms2ger I just updated the pull request with the changes you suggested. How can we also make the layout code make use of this attribute caching?

**Comment by jdm** [comment]
```
failures:
    size_of::size_div
    size_of::size_element
    size_of::size_htmlelement
    size_of::size_span
```
The sizes in the test need to be updated, eg:
```
---- size_of::size_span stdout ----
	thread 'size_of::size_span' panicked at 'Your changes have increased the stack size of commonly used DOM struct HTMLSpanElement from 312 to 336. These structs are present in large quantities in the DOM, and increasing the size may dramatically affect our memory footprint. Please consider choosing a design which avoids this increase. If you feel that the increase is necessary, update to the new size in script/tests.rs.', /Users/servo/buildbot/slave/mac2/build/tests/unit/script/size_of.rs:46
```

**Comment by asabil** [comment]
Pushed a fix

**Comment by Ms2ger** [comment]
lgtm, squash the three commits, please?

**Comment by asabil** [comment]
Squashed

**Comment by Ms2ger** [comment]
Thank you!

@bors-servo r+

**Comment by Ms2ger** [comment]
Those failures look relevant, but I still don't see the bug.

**Comment by nox** [comment]
Superseded by #7068.

**Pull request closed without merging**